Add script to auto-generate chain id -> rpc endpoints mapping files

derekpierre opened the pull request:
- Add script to auto-generate json endpoint mapping files for any domain (`lynx`, `tapir`, `mainnet`)
   - The script runs a health check on each rpc endpoint checking chain id and drift of current block time
   - The script runs using asyncio to speed up processing
- Expand the chain ids supported by `lynx`
- Added pre-commit hooks and linting

derekpierre committed:
Create scripts folder and add sorting script to it.
Updated sorting script to use pathlib.Path relative to the location of the script.

derekpierre committed:
Rename sort_json.py script to sort_chain_endpoints_files.py.

derekpierre committed:
Add script to auto-generate lynx json file with chain -> rpc endpoints mapping for relevant lynx chain ids.
The script verifies the rpc endpoint for both chain id, and limited max block time drift.
Expanded the chain ids configured for lynx.
Generate a lynx.json file using the script.

theref commented on scripts/generate_chain_endpoints_mapping.py:
```diff
@@ -0,0 +1,244 @@
+    11155420,  # OP Sepolia Testnet
+}
+
+EXTRA_KNOWN_RPC_ENDPOINTS = {
```
love that you'd added drpc endpoints here. Similary I think we should add some from https://www.nodies.app/ which is built on top of POKT (similar to dRPC) - although, as I type this, I can't remember if pokt endpoints are compatible with TACo...

`https://ethereum-sepolia-public.nodies.app`

derekpierre replied:
✔️

derekpierre replied:
> I can't remember if pokt endpoints are compatible with TACo...

There was an issue before, but it was fixed via https://github.com/nucypher/nucypher/pull/3491 for nodes to be agnostic to RPC endpoints.

theref reviewed (commented):
Given that we're starting to add some proper scripts in here, how do you feel about a `pre-commit` file that runs black/ruff/etc and also runs the `sort_chain_endpoints_files.py`

derekpierre committed:
Rename scripts and add scripts section to Pipfile.

derekpierre committed:
Add public rpc endpoints from nodies.app to the extra rpc endpoints dictionary used for generating endpoint mappings.

derekpierre committed:
Add pre-commit hooks and run them on files in repos.

derekpierre committed:
Add CI workflow files for darker and ruff.
Add .gitignore file.

derekpierre committed:
Add pre-commit hook to ensure that json files are always sorted.

theref commented on .github/workflows/darker.yml:
```diff
@@ -0,0 +1,21 @@
+name: 🧹 Darker
```
We can just use black - we use darker on `nucypher` so that we don't end up with one massive formatting commit

derekpierre replied:
✔️

derekpierre committed:
Update sort_endpoint_mapping script to use click.
Created common utils.py module for the script to use.
Update local pre-commit hook for sorting JSON files to always run.

derekpierre committed:
Update endpoint mapping generation script to use click, and take domain as an optional parameter.
Update script to also generate endpoint mappings for tapir and mainnet domains.
Remove chain id for Filecoin Butterly testnet and use Filecoin Calibration testnet instead.
Raise exception whenever one of the expected chains has no (healthy) endpoints that can be used.
Raise exception whenever chain id of the endpoint does not match expected chain id configuration.

derekpierre committed:
Use black instead of darker.
Update relevant files that yielded linting errors.

derekpierre committed:
Sorting of mapping should also ensure that there are no repeats of rpc endpoints.

derekpierre committed:
Update known RPC endpoints for Canto Testnet and Celo.
Clean up health check messages for endpoint mapping generation script.
Regenerate lynx.json.

derekpierre committed:
Add README.
